scanner/src/main/kotlin/experimental/ExperimentalScanner.kt

```diff
@@ -156,30 +161,38 @@ class ExperimentalScanner(
         return results + issueResults
     }
 
-    private fun resolvePackageProvenances(controller: ScanController) {
+    private suspend fun resolvePackageProvenances(controller: ScanController) {
         logger.info { "Resolving provenance for ${controller.packages.size} package(s)." }
 
-        val duration = measureTime {
-            controller.packages.forEach { pkg ->
-                runCatching {
-                    packageProvenanceResolver.resolveProvenance(pkg, downloaderConfig.sourceCodeOrigins)
-                }.onSuccess { provenance ->
-                    controller.addPackageProvenance(pkg.id, provenance)
-                }.onFailure {
-                    controller.addProvenanceResolutionIssue(
-                        pkg.id,
-                        OrtIssue(
-                            source = TOOL_NAME,
-                            severity = Severity.ERROR,
-                            message = "Could not resolve provenance for package '${pkg.id.toCoordinates()}': " +
-                                    it.collectMessages()
+        withContext(Dispatchers.IO) {
+            val duration = measureTime {
+                controller.packages.map { pkg ->
+                    async {
+                        pkg to runCatching {
+                            packageProvenanceResolver.resolveProvenance(pkg, downloaderConfig.sourceCodeOrigins)
+                        }
+                    }
+                }.awaitAll().forEach { (pkg, result) ->
+                    result.onSuccess { provenance ->
```

Comment:
I think it should be safe because no other function is interacting with the controller at the same time, but I have applied the change because it also better isolates the asynchronous code and reduces nesting.

Reply:
It is not only an issue with potential data corruption due to concurrent access, but also with safe publication of data. See for instance https://liyanxu.blog/2018/05/19/java-thread-safe-problems/.